### PR TITLE
Release script hot fix changes + "coldfix"

### PR DESCRIPTION
**Pull request by THISISDINOSAUR**
Task/Issue URL: https://app.asana.com/0/414235014887631/1204947064806401/f
Tech Design URL:
CC:

**Description**:
Changes the release script to support changes to the hot fixing process, and adds a new "cold fix" process (i.e. a process for doing release fixes before the release has gone out which doesn't increment the release number).

The general idea behind the hotfix changes are that we should only ever merge in the minimum amount to get stable, i.e. only the code that fixes the bug. It also codifies that we always increment the patch number for a hotfix.

For a hotfix:
As input:
- Takes the version number of *the release you want to hotfix* and then handles the versioning of the new release automatically (by incrementing the patch number).
- Takes a branch name with the fix (which should contain only the fix)

It then:
- Checks the tag exists, and then checks out the tag for the release
- Creates a branch for the hotffix, and then a "changes" branch as with the existing process
- Merges the supplied branch with the fix into this changes branch
- Creates the pull request for the changes branch into the hotfix branch as with the existing process

Perhaps more important is what it does not do:
- It does not update embedded files
- It does not pull in develop

A cold fix is the same, except uses the same version number.

**Steps to test this PR**:
1. Test the script still works for a regular release (i.e. by not supplying -h -c or a fix branch). Check that it still handles the version number currently and updates embedded files as expected
2. Create a small test fix of a release tag (e.g. I have been using elle/release-script-hot-fix-changes-test-fix-branch)
3. Test the script by supplying that version number and that branch with the flag -h. (e.g. `scripts/prepare_release.sh 7.80.0 elle/release-script-hot-fix-changes-test-fix-branch -h`). Check the created PR only contains the updated version number (Which should just be an increment of the patch number), updated release notes, the code changes for the fix (and only the fix), and check that it did not update embedded files.
4. Do the same as with 3, except supply -c instead of -h. The outcome should be identical except the version number should be unchanged, and the build number should instead be incremented as usual. 

**Copy Testing**:

* [ ] Use of correct apostrophes in new copy, ie `’` rather than `'`

**Orientation Testing**:

* [ ] Portrait
* [ ] Landscape

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [ ] iPhone 8
* [ ] iPhone X
* [ ] iPhone 14 Pro
* [ ] iPad

**OS Testing**:

* [ ] iOS 14
* [ ] iOS 15
* [ ] iOS 16

**Theme Testing**:

* [ ] Light theme
* [ ] Dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: scripts/prepare_release.sh [hunk 1]
```diff
@@ -50,11 +51,12 @@ print_usage_and_exit() {
 
 	cat <<- EOF
 	Usage:
-	  $ $(basename "$0") <version> [-h] [-v]
+	  $ $(basename "$0") <version> [banch-with-fix] [-h] [-v]
```
Comment (jaceklyp) [inline scripts/prepare_release.sh]: typo :)

File: scripts/prepare_release.sh [hunk 2]
```diff
@@ -86,7 +105,16 @@ read_command_line_arguments() {
...
 	shift $((OPTIND-1))
 
-	[[ $is_hotfix ]] && branch_name="hotfix" || branch_name="release"
+	if [[ $is_hotfix ]]; then
+		version_to_hotfix=${version}
+		if ! [[ $is_coldfix ]]; then
+			# shellcheck disable=SC2206
+			arrIN=(${version//./ })
+			patch_number=$((arrIN[2]+1))
+			version="${arrIN[0]}.${arrIN[1]}.${patch_number}"
```
Comment (jaceklyp) [inline scripts/prepare_release.sh]: perhaps you can use IFS?

    IFS='.' read -ra arrIN <<< "$version"
    patch_number=$((arrIN[2] + 1))
    version="${arrIN[0]}.${arrIN[1]}.$patch_number"

File: scripts/prepare_release.sh [hunk 3]
```diff
@@ -50,11 +51,12 @@ print_usage_and_exit() {
 
 	cat <<- EOF
 	Usage:
-	  $ $(basename "$0") <version> [-h] [-v]
+	  $ $(basename "$0") <version> [banch-with-fix] [-h] [-v]
 	  Current version: $(cut -d' ' -f3 < "${base_dir}/Configuration/Version.xcconfig")
 
 	Options:
-	  -h         Make hotfix release
+	  -h         Make hotfix release. Requires the version to be the one to hotfix, and a branch with the fix as the second parameter
```
Comment (jaceklyp) [inline scripts/prepare_release.sh:58]: do we provide any error info if someone forgets to provide a branch name?

**Review by jaceklyp — CHANGES_REQUESTED** [review]

File: scripts/prepare_release.sh (hunk 3 above)
Comment (THISISDINOSAUR) [reply scripts/prepare_release.sh:58]: No, it fails silently (although at least in a fairly obvious way:
"Merging fix branch ... %                                                     
ellesullivanwork@Elles-MBP iOS %"

Is there an easy way to do it?

Comment (THISISDINOSAUR) [reply scripts/prepare_release.sh:58]: I've added a check based on the number of arguments:

"Usage:
 ...
💥 Error: Wrong number of arguments. Did you specify a fix branch?"

**Review by jaceklyp — APPROVED** [review]

**Pull request merged**